File: php_backend/shared.php
<?php

require_once __DIR__ . '/../func-proxy.php';

use PhpProxyHunter\CoreDB;

// Disallow access to this file directly
if (basename(__FILE__) == basename($_SERVER['SCRIPT_FILENAME'])) {
  http_response_code(403);
  exit('Direct access not permitted.');
}

// Always attempt to load .env from project root if Dotenv is available.
if (class_exists('Dotenv\\Dotenv')) {
  try {
    $dotenv = Dotenv\Dotenv::createUnsafeImmutable(dirname(__DIR__, 1));
    $dotenv->load();
    // For compatibility with libraries using getenv()
    foreach ($_ENV as $key => $value) {
      putenv("$key=$value");
    }
  } catch (Dotenv\Exception\InvalidPathException $e) {
    // .env not found — continue with environment variables from server/CI.
    if (is_cli()) {
      echo '[dotenv] .env file not found, skipping load.' . PHP_EOL;
    }
  } catch (Throwable $e) {
    // Any other dotenv-related error should not be fatal for scripts.
    // Log to stdout for CI visibility.
    if (is_cli()) {
      echo '[dotenv] failed to load: ' . $e->getMessage() . PHP_EOL;
    }
  }
}

try {
  try {
    $core_db = init_local_database();
  } catch (\Throwable $localDbError) {
    // If local DB initialization fails, attempt to initialize production DB as a fallback.
    error_log('[shared.php] Local DB initialization failed: ' . $localDbError->getMessage());
    if (is_cli()) {
      echo '[shared.php] Local DB initialization failed: ' . $localDbError->getMessage() . PHP_EOL;
    }
    $core_db = init_production_database();
  }
} catch (\Throwable $th) {
  // Log the error so it's visible in logs and CLI output instead of silently
  // swallowing the exception which leads to null DB helpers later.
  error_log('[shared.php] CoreDB initialization failed: ' . $th->getMessage());
  if (function_exists('is_cli') && is_cli()) {
    echo '[shared.php] CoreDB initialization failed: ' . $th->getMessage() . PHP_EOL;
    // Exit early in CLI scripts to avoid passing a broken/null $core_db
    exit(1);
  } else {
    // In web contexts, we can choose to continue with $core_db as null and
    // handle it gracefully in the application (e.g., show an error message).
    $core_db = null;
  }
}

/** @var \PhpProxyHunter\UserDB $user_db */
$user_db = $core_db->user_db ?? null;
/** @var \PhpProxyHunter\ProxyDB $proxy_db */
$proxy_db = $core_db->proxy_db ?? null;
/** @var \PhpProxyHunter\ActivityLog $log_db */
$log_db = $core_db->log_db ?? null;

/**
 * Get current authenticated user data from the database.
 *
 * @return array|null Returns user data array if found, null otherwise
 */
function getCurrentUserData() {
  global $user_db;
  $email = !is_cli() ? ($_SESSION['authenticated_email'] ?? '') : '';
  $user  = $user_db->select($email);
  return !empty($user) ? $user : null;
}

/**
 * Refresh database connections by re-instantiating CoreDB and its DB wrappers.
 *
 * This function performs the following steps:
 * - Declares the global variables that hold DB configuration and connection objects.
 * - Unsets existing connection variables ($core_db, $user_db, $proxy_db, $log_db) to
 *   ensure old connections are released.
 * - Invokes garbage collection to free resources immediately.
 * - Creates a new CoreDB instance with the current configuration and re-assigns
 *   the wrapper properties ($user_db, $proxy_db, $log_db) from the new CoreDB.
 *
 * Globals used:
 * @global string $dbFile Path to SQLite file or ignored for MySQL
 * @global string $dbHost Database host
 * @global string $dbName Database name
 * @global string $dbUser Database username
 * @global string $dbPass Database password
 * @global string $dbType Database type identifier ('sqlite' or 'mysql')
 * @global \PhpProxyHunter\CoreDB $core_db The CoreDB instance to re-create
 * @global \PhpProxyHunter\UserDB|null $user_db The user DB wrapper (may be uninitialized)
 * @global \PhpProxyHunter\ProxyDB|null $proxy_db The proxy DB wrapper (may be uninitialized)
 * @global \PhpProxyHunter\ActivityLog|null $log_db The activity log wrapper (may be uninitialized)
 * @param bool $useFallbackProduction Whether to retry with production DB credentials
 *   when MySQL returns a connection refused error (SQLSTATE[HY000] [2002]).
 *
 * @return array{
 *   core_db:\PhpProxyHunter\CoreDB,
 *   user_db:\PhpProxyHunter\UserDB,
 *   proxy_db:\PhpProxyHunter\ProxyDB,
 *   log_db:\PhpProxyHunter\ActivityLog
 * }|null Returns the newly created CoreDB instance and its DB wrappers, or null
 *   if connection setup fails.
 */
function refreshDbConnections(bool $useFallbackProduction = false): ?array {
  global $dbFile, $dbHost, $dbName, $dbUser, $dbPass, $dbType, $core_db, $user_db, $proxy_db, $log_db;

  // Clean up existing DB connections to avoid conflicts
  unset($core_db, $user_db, $proxy_db, $log_db);
  gc_collect_cycles();

  try {
    $core_db = new CoreDB(
      $dbFile,
      $dbHost,
      $dbName,
      $dbUser,
      $dbPass,
      false,
      $dbType
    );
  } catch (\Throwable $th) {
    $errorMessage        = $th->getMessage();
    $isConnectionRefused = str_contains($errorMessage, 'SQLSTATE[HY000] [2002]')
      && str_contains(strtolower($errorMessage), 'actively refused');

    if (!($useFallbackProduction && $isConnectionRefused)) {
      // Handle connection errors gracefully
      echo 'Error refreshing DB connections: ' . $errorMessage . PHP_EOL;
      return null;
    }

    // Retry with production credentials when explicitly requested.
    $productionDbName = $_ENV['MYSQL_DBNAME_PRODUCTION'] ?? getenv('MYSQL_DBNAME_PRODUCTION') ?: ($_ENV['MYSQL_DBNAME'] ?? getenv('MYSQL_DBNAME'));
    $productionDbUser = $_ENV['MYSQL_USER_PRODUCTION']   ?? getenv('MYSQL_USER_PRODUCTION') ?: ($_ENV['MYSQL_USER'] ?? getenv('MYSQL_USER'));
    $productionDbPass = $_ENV['MYSQL_PASS_PRODUCTION']   ?? getenv('MYSQL_PASS_PRODUCTION') ?: ($_ENV['MYSQL_PASS'] ?? getenv('MYSQL_PASS'));
    $productionDbHost = $_ENV['MYSQL_HOST_PRODUCTION']   ?? getenv('MYSQL_HOST_PRODUCTION') ?: ($_ENV['MYSQL_HOST'] ?? getenv('MYSQL_HOST'));

    try {
      $core_db = new CoreDB(
        $dbFile,
        $productionDbHost,
        $productionDbName,
        $productionDbUser,
        $productionDbPass,
        false,
        'mysql'
      );

      // Persist fallback credentials for subsequent reconnect attempts.
      $dbHost = $productionDbHost;
      $dbName = $productionDbName;
      $dbUser = $productionDbUser;
      $dbPass = $productionDbPass;
      $dbType = 'mysql';
    } catch (\Throwable $fallbackTh) {
      echo 'Error refreshing DB connections: ' . $fallbackTh->getMessage() . PHP_EOL;
      return null;
    }
  }
  /** @var \PhpProxyHunter\UserDB $user_db */
  $user_db = $core_db->user_db;
  /** @var \PhpProxyHunter\ProxyDB $proxy_db */
  $proxy_db = $core_db->proxy_db;
  /** @var \PhpProxyHunter\ActivityLog $log_db */
  $log_db = $core_db->log_db;

  return ['core_db' => $core_db, 'user_db' => $user_db, 'proxy_db' => $proxy_db, 'log_db' => $log_db];
}

/**
 * Initialize a production CoreDB instance using production environment
 * credentials (falls back to standard env vars when needed).
 *
 * This creates and returns a configured `CoreDB` object. Callers should
 * handle any exceptions thrown during connection initialization.
 *
 * @return \PhpProxyHunter\CoreDB
 */
function init_production_database(): CoreDB {
  $dbName = is_debug() ? 'php_proxy_hunter_test' : ($_ENV['MYSQL_DBNAME'] ?? getenv('MYSQL_DBNAME'));
  $dbUser = $_ENV['MYSQL_USER_PRODUCTION'] ?? getenv('MYSQL_USER_PRODUCTION');
  $dbPass = $_ENV['MYSQL_PASS_PRODUCTION'] ?? getenv('MYSQL_PASS_PRODUCTION');
  $dbHost = $_ENV['MYSQL_HOST_PRODUCTION'] ?? getenv('MYSQL_HOST_PRODUCTION');
  $dbFile = is_debug() ? get_project_root('tmp/database_test.sqlite') : get_project_root('src/database.sqlite');
  // Github CI uses SQLite for testing to avoid needing a MySQL service
  $dbType = is_github_ci() ? 'sqlite' : 'mysql';
  return new CoreDB(
    $dbFile,
    $dbHost,
    $dbName,
    $dbUser,
    $dbPass,
    false,
    $dbType
  );
}

/**
 * Initialize a local CoreDB instance using local environment credentials.
 *
 * This creates and returns a configured `CoreDB` object for local or
 * development usage. Callers should handle any exceptions thrown during
 * connection initialization.
 *
 * @return \PhpProxyHunter\CoreDB
 */
function init_local_database(): CoreDB {
  $dbName = is_debug() ? 'php_proxy_hunter_test' : ($_ENV['MYSQL_DBNAME'] ?? getenv('MYSQL_DBNAME'));
  $dbUser = $_ENV['MYSQL_USER'] ?? getenv('MYSQL_USER');
  $dbPass = $_ENV['MYSQL_PASS'] ?? getenv('MYSQL_PASS');
  $dbHost = $_ENV['MYSQL_HOST'] ?? getenv('MYSQL_HOST');
  $dbFile = is_debug() ? get_project_root('tmp/database_test.sqlite') : get_project_root('src/database.sqlite');
  // Github CI uses SQLite for testing to avoid needing a MySQL service
  $dbType = is_github_ci() ? 'sqlite' : 'mysql';
  return new CoreDB(
    $dbFile,
    $dbHost,
    $dbName,
    $dbUser,
    $dbPass,
    false,
    $dbType
  );
}

File: src/shared.py
#!/usr/bin/env python3
"""
Shared database initialization and configuration.

Similar to php_backend/shared.php - handles .env loading and database setup.
"""

import os
import sys

# Add parent directory to path
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

# Load environment variables from .env file
try:
    from dotenv import load_dotenv

    env_file = os.path.join(os.path.dirname(__file__), "..", ".env")
    if os.path.exists(env_file):
        load_dotenv(env_file)
except ImportError:
    pass
except Exception as e:
    print(f"Error loading .env: {e}")

from src.ProxyDB import ProxyDB
from src.func import get_relative_path


def init_db(db_type: str = "sqlite", production: bool = False):
    """
    Initialize ProxyDB with environment configuration.

    Args:
        db_type (str): Database type - 'sqlite' or 'mysql' (default: sqlite)
        production (bool): When True, force use of production MySQL configuration.

    Environment variables:
    - MYSQL_HOST: MySQL host (default: localhost)
    - MYSQL_DBNAME: MySQL database name (default: php_proxy_hunter)
    - MYSQL_USER: MySQL user (default: root)
    - MYSQL_PASS: MySQL password (default: empty)
    - MYSQL_HOST_PRODUCTION: Production host
    - MYSQL_USER_PRODUCTION: Production user
    - MYSQL_PASS_PRODUCTION: Production password

    Returns:
        ProxyDB: Initialized database instance
    """
    if production:
        db_type = "mysql"

    db_type = db_type.lower()

    if db_type == "mysql":
        # MySQL configuration
        if production:
            db_name = os.getenv("MYSQL_DBNAME", "php_proxy_hunter")
        else:
            db_name = os.getenv("MYSQL_DBNAME", "php_proxy_hunter_test")
        if production:
            db_host = os.getenv(
                "MYSQL_HOST_PRODUCTION", os.getenv("MYSQL_HOST", "localhost")
            )
            db_user = os.getenv(
                "MYSQL_USER_PRODUCTION", os.getenv("MYSQL_USER", "root")
            )
            db_pass = os.getenv("MYSQL_PASS_PRODUCTION", os.getenv("MYSQL_PASS", ""))
        else:
            db_host = os.getenv("MYSQL_HOST", "localhost")
            db_user = os.getenv("MYSQL_USER", "root")
            db_pass = os.getenv("MYSQL_PASS", "")

        proxy_db = ProxyDB(
            db_location=db_name,
            start=True,
            db_type="mysql",
            mysql_host=db_host,
            mysql_dbname=db_name,
            mysql_user=db_user,
            mysql_password=db_pass,
        )
    else:
        # SQLite configuration (default)
        db_file = get_relative_path("src/database.sqlite")

        proxy_db = ProxyDB(
            db_location=db_file,
            start=True,
            db_type="sqlite",
        )

    return proxy_db


def init_readonly_db():
    """
    Initialize a read-only ProxyDB instance.

    Returns:
        ProxyDB: Read-only database instance
    """
    proxy_db = ProxyDB(
        start=True,
        db_type="mysql",
        mysql_dbname="myproject",
        mysql_host="23.94.85.180",
        mysql_user="proxyuser",
        mysql_password="proxypassword",
    )

    return proxy_db


if __name__ == "__main__":
    # Specify database type manually here
    # proxy_db = init_db(db_type="mysql")
    # working = proxy_db.get_working_proxies(False, 10)
    # print(working)

    readonly_db = init_readonly_db()
    working = readonly_db.get_working_proxies(False, 10)
    print(working)
